pkg/controller/volume/persistentvolume/pv_controller_base.go

```diff
@@ -94,98 +99,49 @@ func NewController(p ControllerParameters) *PersistentVolumeController {
-			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return p.KubeClient.Core().PersistentVolumes().Watch(options)
-			},
-		}
-	}
-	controller.volumeSource = volumeSource
-
-	claimSource := p.ClaimSource
-	if claimSource == nil {
-		claimSource = &cache.ListWatch{
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				return p.KubeClient.Core().PersistentVolumeClaims(metav1.NamespaceAll).List(options)
-			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return p.KubeClient.Core().PersistentVolumeClaims(metav1.NamespaceAll).Watch(options)
-			},
-		}
-	}
-	controller.claimSource = claimSource
-
-	classSource := p.ClassSource
-	if classSource == nil {
-		classSource = &cache.ListWatch{
-			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				return p.KubeClient.Storage().StorageClasses().List(options)
-			},
-			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-				return p.KubeClient.Storage().StorageClasses().Watch(options)
-			},
-		}
-	}
-	controller.classSource = classSource
-
-	controller.volumeInformer, controller.volumeController = cache.NewIndexerInformer(
-		volumeSource,
-		&v1.PersistentVolume{},
-		p.SyncPeriod,
+	p.VolumeInformer.Informer().AddEventHandlerWithResyncPeriod(
 		cache.ResourceEventHandlerFuncs{
 			AddFunc:    func(obj interface{}) { controller.enqueueWork(controller.volumeQueue, obj) },
 			UpdateFunc: func(oldObj, newObj interface{}) { controller.enqueueWork(controller.volumeQueue, newObj) },
 			DeleteFunc: func(obj interface{}) { controller.enqueueWork(controller.volumeQueue, obj) },
 		},
-		cache.Indexers{"accessmodes": accessModesIndexFunc},
-	)
-	controller.claimInformer, controller.claimController = cache.NewInformer(
-		claimSource,
-		&v1.PersistentVolumeClaim{},
 		p.SyncPeriod,
+	)
+	controller.volumeLister = p.VolumeInformer.Lister()
+	controller.volumeListerSynced = p.VolumeInformer.Informer().HasSynced
+
+	p.ClaimInformer.Informer().AddEventHandlerWithResyncPeriod(
 		cache.ResourceEventHandlerFuncs{
 			AddFunc:    func(obj interface{}) { controller.enqueueWork(controller.claimQueue, obj) },
 			UpdateFunc: func(oldObj, newObj interface{}) { controller.enqueueWork(controller.claimQueue, newObj) },
 			DeleteFunc: func(obj interface{}) { controller.enqueueWork(controller.claimQueue, obj) },
 		},
+		p.SyncPeriod,
 	)
+	controller.claimLister = p.ClaimInformer.Lister()
+	controller.claimListerSynced = p.ClaimInformer.Informer().HasSynced
 
 	// This is just a cache of StorageClass instances, no special actions are
 	// needed when a class is created/deleted/updated.
-	controller.classes = cache.NewStore(cache.DeletionHandlingMetaNamespaceKeyFunc)
-	controller.classReflector = cache.NewReflector(
-		classSource,
-		&storage.StorageClass{},
-		controller.classes,
-		p.SyncPeriod,
-	)
+	controller.classes = p.ClassInformer.Informer().GetStore()
```

Comment:
Sure, let me amend my comment slightly. If all you want to do is list/watch a resource and do something in reaction to seeing the updates come through, I'd say use the shared informer, its associated lister, and add your event handlers. Never mutate any object you get back from the lister (DeepCopy it first). Never mutate the informer's store (don't add/update/delete) directly. If you need to do any of those things, don't use a shared informer.